docs: add guide for standalone migration (#49044)
angular/angular: aio/content/guide/standalone-migration.md

As of version 15.2.0, Angular offers a [schematic](guide/schematics) to help project authors convert existing projects to [the new standalone APIs](guide/standalone-components). The schematic aims to transform as much code as possible automatically, but it may require some manual fixes by the project author. Run the schematic with the following command:

<code-example format="shell" language="shell">

ng generate @angular/core:standalone

</code-example>

## Prerequisites

Before using the schematic, please ensure that the project:
1. Is using Angular 15.2.0 or later.
2. Builds without any compilation errors.
3. Is on a clean Git branch and all work is saved.

## Schematic options

| Option              | Details                                                    |
|:---                 |:---                                                        |
| `mode`              | The transformation to perform. See [Migration modes](#migration-modes) below for details on the available options. |
| `path`              | The path to migrate, relative to the project root. You can use this option to migrate sections of your project incrementally. |

## Migrations steps

> The migration process is composed of three steps which means that you'll have to run it multiple times and check manually that the project builds and behaves as expected.

Reviewer: ```suggestion
The migration process is composed of three steps. You'll have to run it multiple times and check manually that the project builds and behaves as expected.
```

# Migrate an existing Angular application to standalone

As of version 15.2.0, Angular offers a [schematic](guide/schematics) to help application authors with converting their existing application to use [the new standalone APIs](guide/standalone-components). The schematic aims to convert as much code as possible automatically, but it may require some manual fixes by the application author. It can be run with the following command:

<code-example format="shell" language="shell">

ng generate @angular/core:standalone

</code-example>

> ## Prerequisites

Author: I agree that it's weird and I was considering whether to put it after the prerequisites. My thinking with leaving it at the top was that a lot of people don't really read through the guide, but they go looking for the command they need to run. I wanted the command to be as visible as possible.

<header>Note</header>

It is expected that the application author will have to perform some manual fixes after each of the steps. While the schematic tries to preserve the application in a working state after each step, there may be some edge cases that haven't been accounted for. Furthermore, because it has to generate new code, the new code's formatting may not match your existing formatting rules.

</div>

While the migration steps can be run in any order, it is expected that they'll be run in the order listed below. It is also recommended to verify that your application builds and runs between each step. An example migration could look as follows:
1. `ng generate @angular/core:standalone`.
2. Select the "Convert all components, directives and pipes to standalone" option.
3. Verify that the application works and commit the changes.
4. `ng generate @angular/core:standalone`.
5. Select the "Remove unnecessary NgModule classes" option.
6. Verify that the application works and commit the changes.
7. `ng generate @angular/core:standalone`.
8. Select the "Bootstrap the application using standalone APIs" option.
9. Verify that the application works and commit the changes.

> 10. Run your linting and formatting checks, and fix any failures. Commit the result.

Author: Changed although I also wanted to emphasize that users should test between each step.

## After the migration

Congratulations, your application has been converted to standalone 🎉. These are some optional follow-up steps you may want to take now:
* Find and remove any remaining NgModules: since the ["Remove unnecessary NgModules" step](#remove-unnecessary-ngmodules) isn't able to remove all modules automatically, you may have to remove the remaining ones yourself.
* Run the project's unit tests and fix any failures.
* Run a automatic code formatting, if the project has one.
* Run a linter over your project and fix any warnings.

## Migration modes
The migration is made up the following modes that are intended to be run in the order they are listed in:
1. Convert declarations to standalone.
2. Remove unnecessary NgModules.
3. Switch to standalone bootstrapping API.

### Convert declarations to standalone

In this mode, the migration will find all of the components, directives and pipes, and convert them to standalone by setting `standalone: true` and adding any dependencies to the `imports` array.

<div class="callout is-helpful">

> NgModules which bootstrap a component are explicitly ignored in this step, because they are likely to be root modules and they would have to be bootstrapped using `bootstrapApplication` instead of `bootstrapModule`. Their declarations will be converted automatically as a part of the ["Switch to standalone bootstrapping API"](#switch-to-standalone-bootstrapping-api) step.
> 

Reviewer: ```suggestion
The schematic ignores NgModules which bootstrap a component during this step because they are likely root modules used by `bootstrapModule` rather than the standalone-compatible `bootstrapApplication`. The schematic converts these declarations automatically as a part of later the ["Switch to standalone bootstrapping API"](#switch-to-standalone-bootstrapping-api) step.
```

</div>

**Before:**
```typescript
// shared.module.ts
@NgModule({
  imports: [CommonModule],
  declarations: [GreeterComponent],
  exports: [GreeterComponent]
})
export class SharedModule {}
```

```typescript
// greeter.component.ts
@Component({
  selector: 'greeter',
  template: '<div *ngIf="showGreeting">Hello</div>',
})
export class GreeterComponent {
  showGreeting = true;
}
```

**After:**
```typescript
// shared.module.ts
@NgModule({
  imports: [CommonModule, GreeterComponent],
  exports: [GreeterComponent]
})
export class SharedModule {}
```

```typescript
// greeter.component.ts
@Component({
  selector: 'greeter',
  template: '<div *ngIf="showGreeting">Hello</div>',
  standalone: true,
  imports: [NgIf]
})
export class GreeterComponent {
  showGreeting = true;
}
```

### Remove unnecessary NgModules

> After converting all declarations to standalone, a lot of NgModules won't be necessary anymore. This step identifies such modules and deletes them, including as many references to them, as possible. If a module reference can't be deleted automatically, the migration will leave a TODO
> comment saying `TODO(standalone-migration): clean up removed NgModule reference manually` so that the author can delete it themselves.

Reviewer: ````suggestion
After converting all declarations to standalone, many NgModules can be safely removed. This step deletes such module declarations and as many corresponding references as possible. If the migration cannot delete a reference automatically, it leaves the following TODO comment so that you can delete the NgModule manually:
```ts
// TODO(standalone-migration): clean up removed NgModule reference manually 
```
````

> A module is considered "safe to remove" if it:

Reviewer: ```suggestion
The migration considers a module safe to remove if that module:
```

* Has no `declarations`.
* Has no `providers`.
* Has no `bootstrap` components.
* Has no `imports` that reference a `ModuleWithProviders` symbol or a module that can't be removed.
* Has no class members. Empty constructors are ignored.

**Before:**
```typescript
// importer.module.ts
@NgModule({
  imports: [FooComponent, BarPipe],
  exports: [FooComponent, BarPipe]
})
export class ImporterModule {}
```

**After:**
```typescript
// importer.module.ts
// Does not exist!
```

### Switch to standalone bootstrapping API

> Converts any usages of the old `bootstrapModule` API to the new `bootstrapApplication`, in addition to switching the root component to standalone and deleting the root module. If the root module has any `providers` or `imports`, the migration will attempt to carry over or import as of that code as possible into the entry point file.

Reviewer: ```suggestion
This step converts any usages of  `bootstrapModule` to the new, standalone-based `bootstrapApplication`. It also switches the root component to `standalone: true` and deletes the root NgModule. If the root module has any `providers` or `imports`, the migration attempts to copy as much of that this configuration as possible into the new bootstrap call.
```

**Before:**
```typescript
// ./app/app.module.ts
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';

@NgModule({
  declarations: [AppComponent],
  bootstrap: [AppComponent]
})
export class AppModule {}
```

```typescript
// ./app/app.component.ts
@Component({ selector: 'app', template: 'hello' })
export class AppComponent {}
```

```typescript
// ./main.ts
import { platformBrowser } from '@angular/platform-browser';
import { AppModule } from './app/app.module';

platformBrowser().bootstrapModule(AppModule).catch(e => console.error(e));
```

**After:**
```typescript
// ./app/app.module.ts
// Does not exist!
```

```typescript
// ./app/app.component.ts
@Component({ selector: 'app', template: 'hello', standalone: true })
export class AppComponent {}
```

```typescript
// ./main.ts
import { bootstrapApplication } from '@angular/platform-browser';
import { AppComponent } from './app/app.component';

bootstrapApplication(AppComponent).catch(e => console.error(e));
```

## Common problems
Some common problems that may prevent the schematic from working correctly include:

> * Compilation errors - if the application has compilation errors, Angular won't be able to analyze and migrate it correctly.

Reviewer: ```suggestion
* Compilation errors - if the application has compilation errors, Angular cannot analyze and migrate it correctly.
```

> * Files not included in a tsconfig - the schematic determines which files to migrate by looking at all of the `tsconfig.json` files and resolving the files from there. If you have files that aren't captured by any tsconfig, they won't be part of the migration.

Reviewer: ```suggestion
* Files not included in a tsconfig - the schematic determines which files to migrate by analyzing your project's `tsconfig.json` files. The schematic excludes any files not captured by a tsconfig.
```

> * Code that can't be statically analyzed - static analysis is used by the schematic to understand the application and determine where changes have to be made. If there are places in the metadata of a class where the value can't be analyzed, the migration can decide to skip it.

Reviewer: ```suggestion
* Code that cannot be statically analyzed - the schematic uses static analysis to understand your code and determine where to make changes. The migration may skip any classes with metadata that cannot be statically analyzed at build time.
```

## Limitations

> Due to the size and complexity of the migration, there are some cases that haven't been handled:

Reviewer: ```suggestion
Due to the size and complexity of the migration, there are some cases that the schematic cannot handle:
```

> * `imports` added to components in unit tests might not be entirely correct, because the schematic uses the Angular compiler to analyze the application and unit tests aren't supported because they aren't compiled ahead-of-time.

Reviewer: ```suggestion
* Because unit tests are not ahead-of-time (AoT) compiled, `imports` added to components in unit tests might not be entirely correct.
```

> * The schematic is looking for the direct calls of Angular APIs so any custom wrappers around them won't be recognized. E.g. if there is a custom `customConfigureTestModule` function that wraps `TestBed.configureTestingModule`, components it declares may not be recognized.

Reviewer: ```suggestion
* The schematic relies on direct calls to Angular APIs. The schematic cannot recognize custom wrappers around Angular APIs. For example, if there you define a custom `customConfigureTestModule` function that wraps `TestBed.configureTestingModule`, components it declares may not be recognized.
```
